Sources/MongoSwift/MongoCollection+BulkWrite.swift

```diff
@@ -17,118 +17,387 @@ extension MongoCollection {
+                    upsert: Bool? = nil) {
+            self.filter = filter
+            self.update = update
+            self.options = UpdateModelOptions(arrayFilters: arrayFilters, collation: collation, upsert: upsert)
+        }
 
-    /// A set of filters specifying to which array elements an update should apply.
-    public let arrayFilters: [Document]?
+        /// Adds the `updateOne` operation to a bulk write
+        fileprivate func addToBulkWrite(bulk: OpaquePointer) throws {
+            let opts = try BsonEncoder().encode(self.options)
+            var error = bson_error_t()
 
-    /// Specifies a collation to use.
-    public let collation: Document?
+            guard mongoc_bulk_operation_update_one_with_opts(bulk, filter.data, update.data, opts.data, &error) else {
+                throw MongoError.invalidArgument(message: toErrorString(error))
+            }
+        }
+    }
+
+    /// A model for an `updateMany` operation within a bulk write.
+    public struct UpdateManyModel: DeleteOrUpdateWriteModel {
+        private let filter: Document
+        private let update: Document
+        private let options: UpdateModelOptions
+
+        /**
+         * Create a `updateMany` operation for a bulk write.
+         *
+         * - Parameters:
+         *   - filter: A `Document` representing the match criteria
+         *   - update: A `Document` containing update operators
+         *   - arrayFilters: A set of filters specifying to which array elements an update should apply
+         *   - collation: Specifies a collation to use
+         *   - upsert: When `true`, creates a new document if no document matches the query
+         */
+        public init(filter: Document, update: Document, arrayFilters: [Document]? = nil, collation: Document? = nil,
+                    upsert: Bool? = nil) {
+            self.filter = filter
+            self.update = update
+            self.options = UpdateModelOptions(arrayFilters: arrayFilters, collation: collation, upsert: upsert)
+        }
 
-    /// When `true`, creates a new document if no document matches the query.
-    public let upsert: Bool?
+        /// Adds the `updateMany` operation to a bulk write
+        fileprivate func addToBulkWrite(bulk: OpaquePointer) throws {
+            let opts = try BsonEncoder().encode(self.options)
+            var error = bson_error_t()
+
+            guard mongoc_bulk_operation_update_many_with_opts(bulk, filter.data, update.data, opts.data, &error) else {
+                throw MongoError.invalidArgument(message: toErrorString(error))
+            }
+        }
+    }
 }
 
-/// A model for an `updateMany` command.
-public struct UpdateManyModel: WriteModel {
-    /// A `Document` representing the match criteria.
-    public let filter: Document
+/// A protocol indicating write operations that can be batched together using `MongoCollection.bulkWrite`.
+public protocol WriteModel {}
```

Comment:
per my comment on the test file below: I am a little concerned about having a protocol with zero requirements since we have no way of making it `final`. it feels like bad info to the user that they can do something like you did in your test, i.e. make some type, extend `WriteModel`, and pass it in and get an error. we're actually requiring that you use one of *our* `WriteModel`s, not just any `WriteModel`. 

I have a couple of ideas for refactoring this. 

number 1:
- create a simple public class `BulkWriteOperation` which has an internal pointer to a `mongoc_bulk_operation_t`
- put a method `addToBulkWrite(bulk: BulkWriteOperation) throws` in the protocol

this makes it more clear to users what we *actually* require of a `WriteModel`. and since the `OpaquePointer` would be `internal`, they wouldn't actually be able to see or use it. then it would at least be clear to them that there's a requirement (which they are unable to fulfill), as opposed to looking like they get free protocol conformance a la `Codable`.

this is pretty much the pattern I used with `BsonValue` and the encode/decode methods on it. they use `DocumentStorage` and `DocumentIterator`. users can't access the internals of those types, so the only way they could actually make their custom type implement the protocol is by having it delegate to the protocol methods on existing `BsonValue` types (for example, `Int16` could be made a `BsonValue` by just calling `Int32`'s methods and doing conversions.)  

number 2: 
I've also seen enums w/ associated values suggested as solutions to this kind of thing, i.e. we'd have a like `WriteModel` enum with `.delete(DeleteModel)`, `.updateMany(UpdateManyModel)`, and so on, or something like that. users would then have to wrap their models in cases though which is kinda annoying, but this could be more appropriate if we foresee no cases where someone could write a custom `WriteModel` that could actually work by just calling another `WriteModel` type's `addToBulkWrite`. 

one could argue that `BsonValue` should be an enum too, but that would be a huge refactor and I do think there are some valid use cases for users actually extending other types to implement it, e.g. the `Int16` example.

thoughts?

Reply:
I share your concerns about the WriteModel protocol.

Regarding the second suggestion, an enum seems like it'd only make this more verbose/cumbersome, so I'd prefer to avoid that if possible. These model classes already feel quite clunky.

Let's consider the first suggestion of a BulkWriteOperation to abstract `mongoc_bulk_operation_t` and then having the protocol require a `addToBulkWrite(bulk: BulkWriteOperation)` method. This is similar to what we have now with the `fileprivate` `addToBulkWrite()` method; however, InsertOneModel needs to be able to return the document's `_id` so it differs slightly. Otherwise, I would already have put `addToBulkWrite()` on WriteModel instead of creating DeleteOrUpdateWriteModel.

I suppose I could have `addToBulkWrite()` for all models take in `insertedIds` by reference to modify it as needed. Alternatively, I could have another prep method just on InsertOneModel that internally obtains the `_id` (creating it as necessary), and simply call that in the special case for InsertOneModel (as we already have in `MongoCollection.bulkWrite()`).

That said, if you think it makes sense to have a BulkWriteOperation class to abstract away some of the internals, why don't we replace the model classes with methods that append write operations to the bulk? This would be closer to libmongoc's API, and is also what PHPC does in [`MongoDB\Driver\BulkWrite`](http://php.net/manual/en/class.mongodb-driver-bulkwrite.php). The `BulkWriteOperation.insertOne()` method could take care of recording the `_id` on its own. One difference would be that all of these methods would now throw, because they'd need to encode and also call the `mongoc_bulk_operation_*` functions.

Reply:
yes, I think the enum would probably make things worse.

either of your suggestions for dealing with `_id` sound fine, I don't have strong feelings.
I was initially thinking could even just have `addToBulkWrite` return a `BsonValue?` but that doesn't handle null inserted IDs.... though it seems like we would be able to tell when it's a null
 _id from the call site though, since we would never create that ourselves and would have to be on there from the beginning, but it's kind of messy.

so the reason I initially went with the model-based API was that it's what's in the [CRUD spec](https://github.com/mongodb/specifications/blob/master/source/crud/crud.rst#bulk-write-models). but if we are ok with deviating from that (I am ok with it), then I agree your suggestion to just have methods that both initialize + add the operations  would make things a lot less clunky. 

I guess using this would then look something like this? or feel free to propose modifications to this. 
```swift
let bulk = BulkWriteOperation(onCollection: myColl) 
try bulk.insertOne(... whatever params) 
try bulk.deleteOne(... whatever params)
let result = try bulk.execute()
```
I have a bit of a naming nit with `insertOne` and `deleteOne` etc because what we're really doing is adding an insertOne operation to the bulk op, not calling `insertOne`, and I can see that a user might be confused and never bother calling execute after that since it kinda looks like calling normal `insertOne`. but i know this is consistent with libmongoc/PHP so maybe it's just me that finds the naming confusing... add[operationname] might be clearer, but it's a but clunkier.

Reply:
An alternative to constructing BulkWriteOperation from the collection would be to declare it within the collection (as we do with the models now) so it can inherit the necessary `collectionType` generic. Also, we could keep `MongoCollection.bulkWrite()` as the executive method. In effect, BulkWriteOperation simply becomes a replacement for the requests array we already pass to `MongoCollection.bulkWrite()`, with the difference that it gets built up in multiple method calls. I think that's on par with initializing requests from a literal array where each element is a newly constructed object.

I'll consult with other drivers about their CRUD spec deviations before revising the PR. PHPC doesn't implement the spec at all. PHPLIB does so in [`Collection::bulkWrite()`](https://docs.mongodb.com/php-library/master/reference/method/MongoDBCollection-bulkWrite/index.html), which basically takes requests as a list of structs (to avoid constructing so many objects).

Reply:
I spoke with @jyemin about how the Java driver deals with the model classes and we concluded that it was preferable to keep them in place and not deviate from the spec too much. I created a BulkWriteOperation class, which also allowed me to restructure the WriteModel protocol and make InsertOneModel consistent with other operations.